agent/node.go

```diff
@@ -633,7 +634,23 @@ func (n *Node) runManager(ctx context.Context, securityConfig *ca.SecurityConfig
 			ready = nil
 		}
 
-		err = n.waitRole(ctx, ca.AgentRole)
+		roleChanged := make(chan struct{})
+		go func() {
+			n.waitRole(ctx, ca.AgentRole)
+			roleChanged <- struct{}{}
+		}()
+
+		select {
+		case <-done:
+			// Fail out if m.Run() returns error, otherwise wait for
+			// role change.
+			if runErr != nil {
+				err = runErr
+			} else {
+				<-roleChanged
```

Comment:
@tonistiigi: `Run` is only supposed to return without an error if `Stop` is called. It looks like the only place `Stop` is called is below in this function when there is a role change.